Add token authentication

TimoKramer opened the pull request:
Token authentication in it's basic form works. You can set a token in
the configuration and use this token to authenticate against the API.
There is the option to use dev-mode which authenticates you always
without token. That is good to be used if you don't want to give
people the opportunity to run an instance e.g. from the Azure Button.

It might be helpful for future use cases as well to avoid complicated
setup when only using locally. To avoid widespread problems in public
space dev-mode is not set as default and needs to be activated in the
config.
    
- Closes #16
- adds token authentication
- token can be set via configuration
- dev-mode disables authentication
- automated basic integration tests

kordano committed:
integrate connection into middleware

kordano committed:
add database connection middleware

kordano committed:
adjust to new datahike version, add connection list

kordano committed:
add description

kordano committed:
cleanup connections after shutdown

kordano committed:
use latests Datahike version

grischoun committed:
Adds localhost:4000 as a valid cross-origin as it is used by datahike-frontend.

kordano committed:
add query administration

kordano committed:
Merge branch 'admin-endpoints' of github.com:replikativ/datahike-server into admin-endpoints

kordano committed:
create and connect to dbs on creation

kordano committed:
add db handler

kordano committed:
add db endpoint

kordano committed:
add better db representation

kordano committed:
Merge branch 'master' into admin-endpoints

kordano committed:
add db

TimoKramer committed:
Add basic integration tests

TimoKramer committed:
Format files with cljfmt

- Add corrected formatting for all clj files
- Add CI step to check formatting
- CI fails on bad formatting

TimoKramer committed:
Add token authentication

Token authentication in it's basic form works. You can set a token in
the configuration and use this token to authenticate against the API.
There is the option to use dev-mode which authenticates you always
without token. Authentication is good to be used if you want to give
people the opportunity to run an instance e.g. from the Azure Button.

Dev mode might be helpful for future use cases as well as to avoid
complicated setup when only using locally. To avoid widespread
problems in public space dev-mode is not set as default and needs to
be activated explicitly in the config.

- Closes #16
- adds token authentication
- token can be set via configuration
- dev-mode disables authentication
- automated basic integration tests
- added unittests

TimoKramer commented:
should be merged after #1 #14 #15

TimoKramer commented:
I would add some more tests to test dev-mode

TimoKramer commented:
and an ADR

TimoKramer committed:
Switch to replikativ CircleCI Orb

TimoKramer committed:
Write Architecture Decision Record for authentication decision

kordano commented on project.clj:
```diff
@@ -4,18 +4,22 @@
-                 [io.replikativ/datahike "0.3.2-SNAPSHOT" :exclusions [io.replikativ/konserve]]
+                 [io.replikativ/konserve "0.6.0-SNAPSHOT"]
+                 [io.replikativ/hitchhiker-tree "0.1.7"]
+                 [io.replikativ/datahike "0.3.2-SNAPSHOT" :exclusions [io.replikativ/konserve io.replikativ/hitchhiker-tree]]
```
These dependencies will be removed once Datahike has a version that includes the correct dependency versions.

TimoKramer replied:
Right, they should be removed. currently it does not work without them.

kordano commented on src/clj/datahike_server/middleware.clj:
```diff
@@ -0,0 +1,34 @@
+   Dev mode always authenticates."
+  [handler]
+  (fn [request]
+    (let [dev-mode? (clojure.core/get-in dc/config [:server :dev-mode])]
```
Why do you use `clojure.core` namespacing here for `get-in`?

TimoKramer replied:
I was trying to just use clojure.core because there was a warning that get-in is overwritten somewhere in konserve or something and I thought it is better this way. I'll change it back.

kordano commented on bin/run-integrationtests:
```diff
@@ -0,0 +1,27 @@
+trap teardown EXIT
+
+function setup() {
+    mv resources/config.edn resources/config.edn.tmp | true
```
This throws `mv: cannot stat 'resources/config.edn': No such file or directory`

TimoKramer replied:
will redirect to /dev/null

TimoKramer committed:
Review changes

clojure.core namespace not needed for get-in

redirecting stdout and stderr on integrationtesting
